Macros

Keats opened the pull request:
Also uniformize function call args between macros and filters

Keats commented:
@SergioBenitez @andrelmartins 
Roughly working macro system based on discusssion in #66 

A few interesting things popped up when working on it:

- current system does not allow to use macro in the same file they are defined, we could use something like `{{ self::my_macro() }}` where `self` resolves to the current template file
- not sure whether to have default args for macros
- going to uniformize args in general so they will all be kwargs, including tests which mean I will change the current syntax and remove the `is divisibleby 3` syntax
- i don't think it's going to be possible to have `include` use macro
- the `namespace::` syntax breaks the Jinja2 highlighting but I think it's worth it

andrelmartins commented:
@Keats Sorry for delay in replying, wanted to get a chance to test it out.

The current implementation seems nice enough to use, at least for me.

I'm in favor of adding the "self::macro()" variant, since that opens the door to more complex macro code and to adding one-off macros to simplify repeated code that's specific to a given page.

As for default arguments, given that you moved macros to kwargs, it seems like a useful next step to implement default arguments. But it is not a significant feature in my view, just a nice to have.

andrelmartins commented:
Sorry for delay in replying, wanted to get a chance to test it out.

The current implementation seems nice enough to use, at least for me.

I'm in favor of adding the "self::macro()" variant, since that opens the
door to more complex macro code and to adding one-off macros to simplify
repeated code that's specific to a given page.

As for default arguments, given that you moved macros to kwargs, it seems
like a useful next step to implement default arguments. But it is not a
significant feature in my view, just a nice to have.



On Wed, Oct 19, 2016 at 3:37 AM, Vincent Prouillet <notifications@github.com
> wrote:

> @SergioBenitez <https://github.com/SergioBenitez> @andrelmartins
> <https://github.com/andrelmartins>
> Roughly working macro system based on discusssion in #66
> <https://github.com/Keats/tera/issues/66>
>
> A few interesting things popped up when working on it:
>
>    - current system does not allow to use macro in the same file they are
>    defined, we could use something like {{ self::my_macro() }} where self
>    resolves to the current template file
>    - not sure whether to have default args for macros
>    - going to uniformize args in general so they will all be kwargs,
>    including tests which mean I will change the current syntax and remove the is
>    divisibleby 3 syntax
>    - i don't think it's going to be possible to have include use macro
>    - the namespace:: syntax breaks the Jinja2 highlighting but I think
>    it's worth it
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Keats/tera/pull/69#issuecomment-254694789>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AExZ9BMLXpEqmqWSAh2GOdLOp1P53YLvks5q1YJTgaJpZM4KafXw>
> .
>

Keats commented:
@andrelmartins no worries i'm still in the middle of moving/vacation so I'll finish it up next week.

I guess those macros don't replace your need for `include` though?

andrelmartins commented:
@Keats Correct, I still need the `include` tag as is.

Keats commented:
Added the `self::` syntax but the whole thing needs cleaning and testing before being mergeable

SergioBenitez commented:
@Keats Just had a chance to look at this, albeit only in a cursory way. Looks fantastic! Great job! Happy to give it a more thorough look once you think it's ready for merging.

Keats commented:
Looking at some tests examples, kwargs would be annoying to use with them and `is oneof(a, b, c)` wouldn't really work so probably not putting kwargs for tests

Keats commented:
@SergioBenitez feel free to do a full review when you have time, I don't think I'll change it significantly

SergioBenitez commented on README.md:
```diff
@@ -364,3 +364,30 @@ Escapes a string's HTML. Specifically, it makes these replacements:
 - & is converted to &amp;
 
+### Macros
+Macros are a simple way to re-use template bits. Think of them as functions that you can call
```
It's "reuse", not "re-use".

SergioBenitez commented on README.md:
```diff
@@ -364,3 +364,30 @@ Escapes a string's HTML. Specifically, it makes these replacements:
 
+### Macros
+Macros are a simple way to re-use template bits. Think of them as functions that you can call
+and that returns some text.
```
"and" shouldn't be there -> "that return text".

SergioBenitez commented on README.md:
```diff
@@ -364,3 +364,30 @@ Escapes a string's HTML. Specifically, it makes these replacements:
+Macros are a simple way to re-use template bits. Think of them as functions that you can call
+and that returns some text.
+
+Macros are defined the following way:
```
"the following way" -> "as follows"

SergioBenitez commented on README.md:
````diff
@@ -364,3 +364,30 @@ Escapes a string's HTML. Specifically, it makes these replacements:
+Macros are defined the following way:
+
+```jinja2
+{% macro hello_world(greeting, language) %}
````
This is a bit of a weird example since it won't emit anything if language is not equal to "en". Maybe you can add some "else if"s or maybe just an "else"?

SergioBenitez commented on README.md:
````diff
@@ -364,3 +364,30 @@ Escapes a string's HTML. Specifically, it makes these replacements:
+{% endmacro hello_world %}
+```
+
+In order to be called, the file containing the macros needs to be loaded first in the file using them:
````
The wording here is a bit odd.

SergioBenitez commented on README.md:
````diff
@@ -364,3 +364,30 @@ Escapes a string's HTML. Specifically, it makes these replacements:
+```jinja2
+{{ macros::hello_world(greeting="Hello World", language="en") }}
+```
+Do note that macros, like other functions in Tera, require keyword arguments.
````
Why is this the case? Positional arguments are nice in some cases.

Keats replied:
Experiment. I want to see how that works in practice and I'll change if it feels annoying

SergioBenitez commented on src/errors.rs:
```diff
@@ -6,10 +6,10 @@ pub type TeraResult<T> = Result<T, TeraError>;
+            display("Was expecting block with name `{}` to be closed, but `{}` is closing at line {:?}, column {:?}",
                     expected, found, line_no, col_no)
-            description("unexpected endblock name")
+            description("unexpected endX name")
```
What is "endX"?

SergioBenitez commented on src/errors.rs:
```diff
@@ -19,20 +19,24 @@ quick_error! {
             description("invalid value")
         }
+        TemplateNotFound(name: String) {
+            display("Template `{}` wasn't found", name)
```
"wasn't" -> "was not"

SergioBenitez commented on src/lib.rs:
```diff
@@ -35,7 +31,8 @@ mod testers;
 
 // Template is meant to be used internally only but is exported for test/bench.
-#[doc(hidden)] pub use template::Template;
+#[doc(hidden)]
```
Why not inline this attribute as well? The others were inlined.

Keats replied:
Intellij rust auto format, i'll open an issue on it in a bit for that

SergioBenitez commented on src/parser.rs:
```diff
@@ -104,6 +110,13 @@ impl_rdp! {
+        // TODO: add default arg?
+        macro_param = @{ simple_ident }
+        macro_params = !@{ macro_param ~ ([","] ~ macro_param )* }
+        macro_definition = _{ identifier ~ ["("] ~ macro_params* ~ [")"]}
```
Shouldn't the `*` be a `?`? `macro_params` already does the repetition.

SergioBenitez commented on src/parser.rs:
```diff
@@ -136,29 +149,34 @@ impl_rdp! {
-        endif_tag       = !@{ tag_start ~ ["endif"] ~ tag_end }
-        endfor_tag      = !@{ tag_start ~ ["endfor"] ~ tag_end }
+        include_tag      = !@{ tag_start ~ ["include"] ~ string ~ tag_end }
+        import_macro_tag = !@{ tag_start ~ ["import"] ~ string ~ ["as"] ~ simple_ident ~ tag_end}
```
Why is the "as ..." part always required?

Keats replied:
To have a namespace for the macros fns, see the post by @andrelmartins on https://github.com/Keats/tera/issues/66

SergioBenitez replied:
Right, but why can't there be a default name, like, the name of the file without the extension(s)?

SergioBenitez commented on src/parser.rs:
```diff
@@ -210,6 +228,19 @@ impl_rdp! {
             },
+            (_: import_macro_tag, &tpl_name: string, &name: simple_ident) => {
+                Ok(Some(Node::ImportMacro {
+                    tpl_name: tpl_name.replace("\"", "").to_string(),
```
`tpl_name.trim_matches("\"").to_string()` conveys more information, I think. It's also faster.

SergioBenitez commented on src/parser.rs:
```diff
@@ -912,13 +985,13 @@ mod tests {
     #[test]
     fn test_ast_if_with_test_params() {
-        let parsed_ast = parse(r#"{% if pi is equalto 3.14 %}Hey{% endif %}"#);
+        let parsed_ast = parse(r#"{% if pi is equalto 3.13 %}Hey{% endif %}"#);
```
Why this change?

Keats replied:
to avoid clippy warning about pi

SergioBenitez commented on src/parser.rs:
```diff
@@ -1062,6 +1135,7 @@ mod tests {
 
     #[test]
     fn test_unary_test() {
+        // TODO: remove that syntax
```
Why? It's very convenient.

SergioBenitez commented on src/parser.rs:
```diff
@@ -1078,6 +1152,7 @@ mod tests {
         assert!(parser.end());
     }
 
+    // TODO: remove that syntax
```
This test checks that the syntax there _doesn't_ work.

SergioBenitez commented on src/parser.rs:
```diff
@@ -1132,6 +1205,60 @@ mod tests {
 
     #[test]
+    fn test_ast_macro_definition() {
+        let parsed_ast = parse("{% macro helloworld(greeting) %}Hello{% endmacro helloworld %}");
```
There should probably be tests that use zero parameters and more than one parameter.

SergioBenitez commented on src/parser.rs:
```diff
@@ -1132,6 +1205,60 @@ mod tests {
+
+    #[test]
+    fn test_ast_macro_call() {
+        let parsed_ast = parse("{{ macros::macro1(foo=bar, hey=1+2) }}");
```
Likewise about the arity on the calling side.

SergioBenitez commented on src/render.rs:
```diff
@@ -345,6 +363,71 @@ impl<'a> Renderer<'a> {
+    // TODO: have a look at this method once it is working to clean it up, too much indentation
+    // Probably very inefficient and can be trimmed down
+    fn render_macro(&mut self, call_node: Node) -> TeraResult<String> {
+        match call_node {
```
Use `if let` instead.

Keats replied:
I didn't realise I could use `if let` for enum variant!

SergioBenitez commented on src/render.rs:
```diff
@@ -345,6 +363,71 @@ impl<'a> Renderer<'a> {
+        match call_node {
+            MacroCall {namespace, name, params} => {
+                // Avoid shadowing when matching the definition of macros
+                let call_params = params;
```
Not sure why this is done.

SergioBenitez commented on src/render.rs:
```diff
@@ -345,6 +363,71 @@ impl<'a> Renderer<'a> {
+                        .expect("Open an issue with a template sample please (mention `self namespace macro`)!")
+                        .to_string()
+                } else {
+                    self.macro_namespaces.push(namespace.clone());
```
Do you really need to clone twice?

SergioBenitez commented on src/render.rs:
```diff
@@ -345,6 +363,71 @@ impl<'a> Renderer<'a> {
+                let call_params = params;
+                // To find the real namespace we just look at the last one we got.
+                // If we have self, we take the last one otherwise we push it into the vec
+                let real_namespace = if namespace == "self" {
```
A `match` would be cleaner here.

SergioBenitez commented on src/render.rs:
```diff
@@ -345,6 +363,71 @@ impl<'a> Renderer<'a> {
+                    namespace.clone()
+                };
+                // TODO: avoid cloned() everywhere if possible
+                match self.macros.get(&real_namespace).cloned() {
```
This is very, very hairy, and likely needlessly so. Use the `Option` combinators to make it much more accessible: `map`, `and_then`, etc.

SergioBenitez commented on src/render.rs:
```diff
@@ -356,6 +439,14 @@ impl<'a> Renderer<'a> {
+                let tpl = try!(self.tera.get_template(&tpl_name));
+                self.macros.insert(name.to_string(), tpl.macros.clone());
+                // In theory, the render_node should return TeraResult<Option<String>>
+                // but in practice there's no difference so keeping this hack
```
Except this allocates needlessly. Just do `let x = match node` at the top. Here, `return Ok(None)`. Everywhere else, remove the `Ok(..)`, but keep what's inside. Then add a final `Ok(x)` to the end of the function.

SergioBenitez commented on src/template.rs:
```diff
@@ -81,4 +107,12 @@ mod tests {
     }
+
+    #[test]
+    fn test_can_find_macros() {
```
This is a fairly weak test.

Keats replied:
granted it doesn't test that much but it's still helpful I think

SergioBenitez commented on tests/expected/use_macros.html:
```diff
@@ -0,0 +1,5 @@
+
```
Why is there so much empty space? Probably don't want this.

SergioBenitez commented on tests/render-failures/macros.html:
```diff
@@ -0,0 +1,10 @@
+{% macro hello_world(greeting, language) %}
+    {% if language == "en" %}
```
Same as my comment in the README: this is strange example.

Keats commented:
@SergioBenitez thanks for the comments!
I'm still not sure what to allow with `self`, I had trouble explaning it in the docs which is not really a good sign. I'll try to come up with a better design this week. Another solution would be to ship an initial version macros that doesn't allow macro calls at the beginning

Keats commented:
Updated the code, thanks for the comments @SergioBenitez 
Still need to think through a bit more the `self` and similar things

Keats commented:
> I feel that some interesting test cases are missing. What happens if a macro is recursive? Should Tera detect that or not? How many levels deep can I nest things? Can I pass macro invocations to macros? Can I pass macro names to macros? A lot of questions are unanswered for me.

Added some more tests and details in the README, hopefully the current state is enough for MVP @SergioBenitez

SergioBenitez commented on README.md:
```diff
@@ -364,3 +364,34 @@ Escapes a string's HTML. Specifically, it makes these replacements:
+
+Macros can be called recursively but there is no limit to recursion so make sure you macro ends.
+
+There are a few restrictions to the content inside macros: no macros definitions, includes or blocks.
```
Why can't you do an `include` inside of a macro? That seems useful.

Keats replied:
I was wondering about that and hoped someone would comment on it! I was not sure myself, I'll add it back

SergioBenitez replied:
Maybe add a test to ensure it works as expected?

Keats replied:
There is one that try to define a macro inside a macro (in parser-failure)

SergioBenitez replied:
Sorry, I meant there should be a test that checks that `include` works inside of a macro. That is, if a macro is called that uses `include`, the `include`d content is rendered.

SergioBenitez commented on src/render.rs:
```diff
@@ -345,6 +363,63 @@ impl<'a> Renderer<'a> {
+                        .to_string()
+                },
+                _ => {
+                    // TODO: String doesn't have Copy trait, can we avoid double cloning?
```
Yes. Keep a `Vec<&str>` instead. Or, a `Vec<Rc<String>>`, if you wish.

Keats replied:
Didn't have to deal with lifetime too much yet, but how would you keep the `&str` values alive longer than the `if let` (without using `Rc`)? I tried a couple of approaches but I can't get a lifetime that lives long enough

SergioBenitez commented on src/render.rs:
```diff
@@ -345,6 +363,63 @@ impl<'a> Renderer<'a> {
+                "self" => {
+                    // TODO: handle error if we don't have a namespace
+                    self.macro_namespaces
+                        .last()
```
The logic here is somewhat unsettling. It's not clear to me how `macro_namespaces` gets populated the firt time. IE: How does this _not_ panic when calling a recursive macro? In that case the namespace will be `self`, but won't `macro_namespaces` be blank? If not, how does it get its elements?

SergioBenitez commented on src/render.rs:
```diff
@@ -345,6 +363,63 @@ impl<'a> Renderer<'a> {
+                _ => {
+                    // TODO: String doesn't have Copy trait, can we avoid double cloning?
+                    self.macro_namespaces.push(namespace.clone());
+                    namespace.clone()
```
You don't need to `clone` here at all. Just use `.as_str` to get a pointer to the `String`.

SergioBenitez commented on src/render.rs:
```diff
@@ -345,6 +363,63 @@ impl<'a> Renderer<'a> {
+        if let MacroCall {namespace, name: macro_name, params: call_params} = call_node {
+            // To find the real namespace we just look at the last one we got.
+            // If we have self, we take the last one otherwise we push it into the vec of namespaces
+            let real_namespace = match namespace.as_ref() {
```
Use `as_str` or `&*namespace`.

SergioBenitez commented on src/render.rs:
```diff
@@ -345,6 +363,63 @@ impl<'a> Renderer<'a> {
+            };
+
+            if let Some(Macro {body, params, ..}) = self.macros.get(&real_namespace)
+                .and_then(|m| m.get(&macro_name).cloned()) {
```
Better to bind the right-hand side of this match and then try to the match:

```
let x = self.macro.get(...).and_then(...);
if let Some(...) = x ...
```

SergioBenitez commented on src/render.rs:
```diff
@@ -345,6 +363,63 @@ impl<'a> Renderer<'a> {
+            if let Some(Macro {body, params, ..}) = self.macros.get(&real_namespace)
+                .and_then(|m| m.get(&macro_name).cloned()) {
+                // fail early if we don't have the same number of args
+                let expected_params = params.iter().cloned().collect::<Vec<String>>();
```
Why is there so much cloning going on here? And why the `collect`? It's remarkably inefficient. If you just want to get a count, use `count`.

SergioBenitez commented on src/render.rs:
```diff
@@ -345,6 +363,63 @@ impl<'a> Renderer<'a> {
+                for node in body.get_children() {
+                    output.push_str(&try!(self.render_node(node)));
+                }
+                // Remove the namespace before returning if necessary
```
This comment doesn't help me understand why this is necessary or important or correct.

SergioBenitez commented on src/render.rs:
```diff
@@ -345,6 +363,63 @@ impl<'a> Renderer<'a> {
+            }
+
+        } else {
+            unreachable!()
```
Tell me why: `unreachable!("This is why.")`.

Keats commented:
For 1, there is a test using a factorial macro, do you mean something else?

SergioBenitez commented:
No, that's perfect. Sorry, I missed it.

Keats commented:
How would you do the macro namespace logic? The current one makes sense to me so I'm interested in seeing other methods.

> Right, but why can't there be a default name, like, the name of the file without the extension(s)?
I can't reply to this inline for some reason.  It's an interesting idea I didn't consider before but that is too implicit imo

SergioBenitez commented:
I don't actually understand the namespace logic in your code, though I haven't spent much time looking at it. Can you explain what your code is doing?

I don't think that's too implicit. Almost every language with imports has a default name when the import isn't qualified, and I don't see why it should be any different in Tera.

Keats commented:
Fixed a couple of bugs and added some more comments in the code to explain it, hopefully it's clearer now - I can't find a simpler way myself.

If you have some time i'd love to see how you would do it to remove the allocations

Keats commented on src/render.rs:
```diff
@@ -345,6 +364,81 @@ impl<'a> Renderer<'a> {
+
+            if let Some(Macro {body, params, ..}) = macro_definition {
+                let expected_params = params.iter().cloned().collect::<Vec<String>>();
+                let params_seen = call_params.keys().cloned().collect::<Vec<String>>();
```
I can remove them from there but then if i only collect the args on errors, I run into that error:

```rust
                for (param_name, exp) in call_params.clone() {
                    if !params.contains(&param_name) {
                        let expected_params = params.iter().cloned().collect::<Vec<String>>();
                        let params_seen = call_params.keys().cloned().collect::<Vec<String>>();
                        return Err(MacroCallWrongArgs(macro_name, expected_params, params_seen));
                    }
                    context.insert(param_name.to_string(), try!(self.eval_expression(exp)));
                }
```
Is there a way to iterate on a `HashMap` without moving the value and without `clone()`? I need the args name for the error message

Keats commented:
Pushed a commit with fewer allocations but still one `clone` for the one mentioned in a comment

SergioBenitez commented on src/render.rs:
```diff
@@ -345,6 +364,81 @@ impl<'a> Renderer<'a> {
+                // We need to make a new context for the macro from the arguments given
+                // Return an error if we get some unknown params
+                let mut context = HashMap::new();
+                for (param_name, exp) in call_params.clone() {
```
I'm pretty sure you don't need this `clone`. Just put an `&` before the `call_params`.

SergioBenitez replied:
Ah, that's a bummer. All of the `eval` expression should allow references as well. Like taking in a T: Borrow<parser::Node> instead of just a parser::Node. But, meh.

SergioBenitez commented:
This is okay for now. I think we should work on removing many of those `clone` and `cloned` calls. They're unnecessary, for the most part.

Keats commented:
> This is okay for now. I think we should work on removing many of those clone and cloned calls. They're unnecessary, for the most part.

Tera is getting close to fully featured in my mind (once macro/escape/super are in) so the focus is going to be on things like that soon!
